src/Datadog.Trace/Util/ObjectPool.cs

```diff
@@ -0,0 +1,105 @@
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Datadog.Trace.Util
+{
+    /// <summary>
+    /// Generic object pool
+    /// </summary>
+    /// <typeparam name="TObject">Type of the pool item object</typeparam>
+    /// <typeparam name="TStruct">Type of the allocator</typeparam>
+    internal class ObjectPool<TObject, TStruct>
+        where TObject : class
+        where TStruct : struct, IPoolAllocator<TObject>
+    {
+        private readonly TStruct _allocator = default;
+        private readonly ConcurrentStack<TObject> _stack = new ConcurrentStack<TObject>();
+        private readonly Timer _dropTimer;
+        private TObject _fastItem;
+        private int _maxItems;
+        private int _minItems;
+        private int _count;
+
+        public ObjectPool(int minItems = 5, int maxItems = 250, int? dropFrequencyInMinutes = 5)
+        {
+            _minItems = minItems;
+            _maxItems = maxItems;
+
+            for (int i = 0; i < minItems; i++)
+            {
+                _stack.Push(_allocator.New());
+            }
+
+            if (dropFrequencyInMinutes.HasValue && dropFrequencyInMinutes > 0)
+            {
+                var frequency = TimeSpan.FromMinutes(dropFrequencyInMinutes.Value);
+                _dropTimer = new Timer(DropTimerMethod, this, frequency, frequency);
+            }
+        }
+
+        public TObject Get()
+        {
+            var value = _fastItem;
+            if (value != null && value == Interlocked.CompareExchange(ref _fastItem, null, value))
+            {
+                return value;
+            }
+
+            if (_stack.TryPop(out value))
+            {
+                Interlocked.Decrement(ref _count);
+                return value;
+            }
+
+            return _allocator.New();
+        }
+
+        public void Return(TObject value)
+        {
+            if (value is null)
+            {
+                return;
+            }
+
+            if (_fastItem is null)
+            {
+                _allocator.Clear(value);
+
+                // we don't care if in a multithread scenario if we replace a reference
```

Comment:
Don't we? It means we lose an instance, which means one more allocation down the line

Reply:
Well at first I did not want to add contention but I guess we can do an Interlocked there.

Reply:
fixed here: https://github.com/DataDog/dd-trace-dotnet/pull/829/commits/d6f75a5d226bee869253ce5236047d590fe6675e